In [1]:
import pandas as pd
import numpy as np
from sklearn.model_selection import train_test_split
from sklearn.preprocessing import StandardScaler
from sklearn.tree import DecisionTreeRegressor
from sklearn.svm import SVR
from sklearn.metrics import mean_squared_error, r2_score
from sklearn.model_selection import train_test_split
from sklearn.linear_model import Ridge
from sklearn.ensemble import RandomForestRegressor, GradientBoostingRegressor
from mlxtend.regressor import StackingRegressor

In [2]:
df = pd.read_csv('att3.csv')
df

date  download  location_latitude  location_longitude  \
0     27-12-2023 13:16    14.190          18.490057           73.852375   
1     27-12-2023 13:16     4.740          18.490057           73.852375   
2     27-12-2023 13:16     2.600          18.490057           73.852375   
3     27-12-2023 13:16     6.800          18.490057           73.852375   
4     27-12-2023 13:16    10.910          18.490057           73.852375   
...                ...       ...                ...                 ...   
6178  27-12-2023 13:57   151.708          18.571069           73.773285   
6179  27-12-2023 13:57   133.327          18.571069           73.773285   
6180  27-12-2023 13:57   128.422          18.571069           73.773285   
6181  27-12-2023 13:57   121.529          18.571069           73.773285   
6182  27-12-2023 13:57   131.890          18.571069           73.773285   

     networkProvider        ping         timestamp  upload  download_no_zeros  \
0             AIRTEL   69.840000  27-12-2023 13:16   8.850             14.190   
1             AIRTEL   76.550000  27-12-2023 13:16   2.390              4.740   
2             AIRTEL   84.330000  27-12-2023 13:16   0.650              2.600   
3             AIRTEL   72.110000  27-12-2023 13:16   3.060              6.800   
4             AIRTEL   88.900000  27-12-2023 13:16   0.590             10.910   
...              ...         ...               ...     ...                ...   
6178          AIRTEL   63.030504  27-12-2023 13:57  11.372            151.708   
6179          AIRTEL  104.493536  27-12-2023 13:57   3.761            133.327   
6180          AIRTEL  109.795346  27-12-2023 13:57   9.211            128.422   
6181          AIRTEL   37.786037  27-12-2023 13:57   4.057            121.529   
6182          AIRTEL  193.947341  27-12-2023 13:57  12.108            131.890   

      upload_no_zeros  ping_no_zeros  
0               8.850      69.840000  
1               2.390      76.550000  
2               0.650      84.330000  
3               3.060      72.110000  
4               0.590      88.900000  
...               ...            ...  
6178           11.372      63.030504  
6179            3.761     104.493536  
6180            9.211     109.795346  
6181            4.057      37.786037  
6182           12.108     193.947341  

[6183 rows x 11 columns]

In [3]:
print(df['networkProvider'].unique()) 

['AIRTEL' 'JIO 4G' 'Vi India']


In [4]:
df.isnull().sum()

date                  0
download              0
location_latitude     0
location_longitude    0
networkProvider       0
ping                  0
timestamp             0
upload                0
download_no_zeros     0
upload_no_zeros       0
ping_no_zeros         0
dtype: int64

Bandwidth (in bps) is the maximum rate at which data can be transmitted on a network connection.
Download speed (also in bps) is the actual rate at which data is being received by your device from the network.
Since download speed is typically measured in bytes per second (Bps), while bandwidth is measured in bits per second (bps), you need to convert between the two units. There are 8 bits in a byte. So, to convert from bytes to bits, you multiply by 8.

In [5]:
df['bandwidth'] = df['download'] * 8
df

date  download  location_latitude  location_longitude  \
0     27-12-2023 13:16    14.190          18.490057           73.852375   
1     27-12-2023 13:16     4.740          18.490057           73.852375   
2     27-12-2023 13:16     2.600          18.490057           73.852375   
3     27-12-2023 13:16     6.800          18.490057           73.852375   
4     27-12-2023 13:16    10.910          18.490057           73.852375   
...                ...       ...                ...                 ...   
6178  27-12-2023 13:57   151.708          18.571069           73.773285   
6179  27-12-2023 13:57   133.327          18.571069           73.773285   
6180  27-12-2023 13:57   128.422          18.571069           73.773285   
6181  27-12-2023 13:57   121.529          18.571069           73.773285   
6182  27-12-2023 13:57   131.890          18.571069           73.773285   

     networkProvider        ping         timestamp  upload  download_no_zeros  \
0             AIRTEL   69.840000  27-12-2023 13:16   8.850             14.190   
1             AIRTEL   76.550000  27-12-2023 13:16   2.390              4.740   
2             AIRTEL   84.330000  27-12-2023 13:16   0.650              2.600   
3             AIRTEL   72.110000  27-12-2023 13:16   3.060              6.800   
4             AIRTEL   88.900000  27-12-2023 13:16   0.590             10.910   
...              ...         ...               ...     ...                ...   
6178          AIRTEL   63.030504  27-12-2023 13:57  11.372            151.708   
6179          AIRTEL  104.493536  27-12-2023 13:57   3.761            133.327   
6180          AIRTEL  109.795346  27-12-2023 13:57   9.211            128.422   
6181          AIRTEL   37.786037  27-12-2023 13:57   4.057            121.529   
6182          AIRTEL  193.947341  27-12-2023 13:57  12.108            131.890   

      upload_no_zeros  ping_no_zeros  bandwidth  
0               8.850      69.840000    113.520  
1               2.390      76.550000     37.920  
2               0.650      84.330000     20.800  
3               3.060      72.110000     54.400  
4               0.590      88.900000     87.280  
...               ...            ...        ...  
6178           11.372      63.030504   1213.664  
6179            3.761     104.493536   1066.616  
6180            9.211     109.795346   1027.376  
6181            4.057      37.786037    972.232  
6182           12.108     193.947341   1055.120  

[6183 rows x 12 columns]

In [26]:

input_latitude = 18.571069

input_longitude = 73.773285

In [27]:
# Calculate pairwise distances between input and dataset coordinates
from scipy.spatial.distance import cdist
input_coords = [[input_latitude, input_longitude]]
dataset_coords = df[['location_latitude', 'location_longitude']].values
distances = cdist(input_coords, dataset_coords, metric='euclidean')


In [28]:
# Find rows with coordinates similar to input based on 70% similarity
threshold_distance = 0.7 * distances.max()
similar_rows_indices = [i for i, distance in enumerate(distances[0]) if distance <= threshold_distance]
similar_rows = df.iloc[similar_rows_indices]

In [29]:
# Print the selected similar rows
print("Rows with similar coordinates to input:")
print(similar_rows)

Rows with similar coordinates to input:
                  date  download  location_latitude  location_longitude  \
5451  27-12-2023 13:16     7.000          18.575037           73.761012   
5452  27-12-2023 13:16     7.230          18.575037           73.761012   
5453  27-12-2023 13:16     7.460          18.575037           73.761012   
5454  27-12-2023 13:16     7.690          18.575037           73.761012   
5455  27-12-2023 13:16     7.920          18.575037           73.761012   
...                ...       ...                ...                 ...   
6178  27-12-2023 13:57   151.708          18.571069           73.773285   
6179  27-12-2023 13:57   133.327          18.571069           73.773285   
6180  27-12-2023 13:57   128.422          18.571069           73.773285   
6181  27-12-2023 13:57   121.529          18.571069           73.773285   
6182  27-12-2023 13:57   131.890          18.571069           73.773285   

     networkProvider        ping         timestamp  upload 

In [30]:
# Define features and target
features = ['download_no_zeros', 'location_latitude', 'location_longitude', 'ping_no_zeros', 'upload_no_zeros']
target = 'bandwidth'

# Split data into train and test sets
X_train, X_test, y_train, y_test = train_test_split(similar_rows[features], similar_rows[target], test_size=0.2, random_state=42)


In [31]:
# Feature scaling
scaler = StandardScaler()
X_train_scaled = scaler.fit_transform(X_train)
X_test_scaled = scaler.transform(X_test)


In [32]:
# Initialize models
model1 = RandomForestRegressor()
model2 = GradientBoostingRegressor()
model3 = SVR()
stacking_model = StackingRegressor(regressors=[model1, model2, model3], meta_regressor=Ridge())

# Fit stacking model
stacking_model.fit(X_train_scaled, y_train)


StackingRegressor(meta_regressor=Ridge(),
                  regressors=[RandomForestRegressor(),
                              GradientBoostingRegressor(), SVR()])

In [33]:
# Sample input
sample_input = pd.DataFrame({
    'download_no_zeros': [0],
    'location_latitude': [input_latitude],
    'location_longitude': [input_longitude],
    'ping_no_zeros': [0],       
    'upload_no_zeros': [0]      
})

# Scale sample input
sample_input_scaled = scaler.transform(sample_input)


In [34]:
# Predict bandwidth for sample input
sample_prediction = stacking_model.predict(sample_input_scaled)
#print(f"Predicted bandwidth for latitude={input_latitude}, longitude={input_longitude}: {sample_prediction[0]}")


In [35]:
# Initialize variables for best provider and best prediction
best_provider = None
best_prediction = float('-inf')

# Iterate over unique network providers in similar rows
for provider in similar_rows['networkProvider'].unique():
    provider_data = similar_rows[similar_rows['networkProvider'] == provider]
    X_provider = provider_data[features]
    X_provider_scaled = scaler.transform(X_provider)

    provider_predictions = stacking_model.predict(X_provider_scaled)
    mean_prediction = provider_predictions.mean()
    
    if mean_prediction > best_prediction:
        best_provider = provider
        best_prediction = mean_prediction
print("Baner -")
print(f"The best provider for latitude={input_latitude}, longitude={input_longitude} is: {best_provider}")


Baner -
The best provider for latitude=18.571069, longitude=73.773285 is: AIRTEL


Baner - 

input_latitude = 18.571069

input_longitude = 73.773285


PVG

input_latitude = 18.490057

input_longitude = 73.8523753

Kothrud - 

input_latitude = 18.50475181

input_longitude = 73.82671298